src/main/java/org/jenkinsci/plugins/ssegateway/EventHistoryStore.java

```diff
@@ -102,9 +102,19 @@ public static void store(@Nonnull Message message) {
             String channelName = message.getChannelName();
             String eventUUID = message.getEventUUID();
             File channelDir = getChannelDir(channelName);
-            File eventFile = new File(channelDir, eventUUID + ".json");
+            
+            // We write to an intermediate file and then do a rename. This should
+            // lower the chances of an EventDispacther (or other) attempting to
+            // read from the file before it is fully written, as the rename should be
+            // considerably "more" (depending on the platform) atomic on most platforms.
+            
+            File writeEventFile = new File(channelDir, eventUUID + "_WRITE.json");
+            File readEventFile = new File(channelDir, eventUUID + ".json");
         
-            FileUtils.writeStringToFile(eventFile, message.toJSON(), "UTF-8");
+            FileUtils.writeStringToFile(writeEventFile, message.toJSON(), "UTF-8");
+            if (!writeEventFile.renameTo(readEventFile)) {
```

Comment:
@tfennelly Thanks. Maybe its not clear, at least to me. So store(message) is called from multiple threads, but each thread has unique message, that is message.getEventUUID() is unique for each message resulting in unique file name?